engines/instance_verification/lib/instance_verification/engine.rb

```diff
@@ -22,6 +22,73 @@ def self.write_cache_file(cache_dir, cache_key)
     FileUtils.touch(File.join(cache_dir, cache_key))
   end
 
+  def self.verify_instance(request, logger, system)
+    return false unless request.headers['X-Instance-Data']
+
+    instance_data = Base64.decode64(request.headers['X-Instance-Data'].to_s)
+    base_product = system.products.find_by(product_type: 'base')
+    return false unless base_product
+
+    # check the cache for the system (20 min)
+    cache_path = File.join(
+      Rails.application.config.repo_cache_dir,
+      InstanceVerification.build_cache_key(request.remote_ip, system.login, base_product_id: base_product.id)
+    )
+    if File.exist?(cache_path)
+      # only update registry cache key
+      InstanceVerification.update_cache(request.remote_ip, system.login, nil, registry: true)
+      return true
+    end
+
+    verification_provider = InstanceVerification.provider.new(
+      logger,
+      request,
+      base_product.attributes.symbolize_keys.slice(:identifier, :version, :arch, :release_type),
+      instance_data
+    )
+
+    is_valid = verification_provider.instance_valid?
+    # update repository and registry cache
+    InstanceVerification.update_cache(request.remote_ip, system.login, base_product.id)
+    is_valid
+  rescue InstanceVerification::Exception => e
+    message = ''
+    if system.proxy_byos
+      result = SccProxy.scc_check_subscription_expiration(request.headers, system.login, system.system_token, logger)
+      if result[:is_active]
+        InstanceVerification.update_cache(request.remote_ip, system.login, base_product.id)
+        return true
+      end
+
+      message = result[:message]
+    else
+      message = e.message
+    end
+    details = [ "System login: #{system.login}", "IP: #{request.remote_ip}" ]
+    details << "Instance ID: #{verification_provider.instance_id}" if verification_provider.instance_id
+    details << "Billing info: #{verification_provider.instance_billing_info}" if verification_provider.instance_billing_info
+
+    ZypperAuth.auth_logger.info <<~LOGMSG
+    Access to the repos denied: #{message}
+      #{details.join(', ')}
+    LOGMSG
+    false
+  rescue StandardError => e
+    logger.error('Unexpected instance verification error has occurred:')
+    logger.error(e.message)
+    logger.error("System login: #{system.login}, IP: #{request.remote_ip}")
+    logger.error('Backtrace:')
+    logger.error(e.backtrace)
+    false
+  end
+
+  def self.build_cache_key(remote_ip, login, base_product_id: nil)
```

Comment:
This seems more complicated than it needs to be. Is there any reason something simpler couldn't be used, since there's no check on the assigned values anyway?

```
def self.build_cache_key(*args)
  args.compact.join('-')
end
```